File: platform/windows/MihomoMeter.Windows.Core/Application/SharedCoreTrafficRouter.cs
using System.Runtime.InteropServices;

namespace MihomoMeter.Windows.Core.Application;

public enum SharedCoreTrafficRouteSource
{
    SharedPrimary,
    NativeFallback,
}

public enum SharedCoreTrafficRouteStatus
{
    Matched,
    Succeeded,
    AbiMismatch,
    NativeCallFailed,
    UnexpectedResult,
    Mismatch,
    UnknownFailure,
}

public readonly record struct SharedCoreTrafficRouteResult(
    string Text,
    SharedCoreTrafficRouteSource Source,
    SharedCoreTrafficShadowStatus Status);

public readonly record struct SharedCoreTrafficLazyRouteResult(
    string Text,
    SharedCoreTrafficRouteSource Source,
    SharedCoreTrafficRouteStatus Status);

internal static class SharedCoreTrafficRouter
{
    public static SharedCoreTrafficRouteResult Route(
        ulong bytes,
        string nativeText,
        SharedCoreTrafficFormat format)
    {
        return Route(
            bytes,
            nativeText,
            format,
            () => MihomoMeterSharedCore.AbiVersion,
            MihomoMeterSharedCore.ScaleTraffic);
    }

    internal static SharedCoreTrafficRouteResult Route(
        ulong bytes,
        string nativeText,
        SharedCoreTrafficFormat format,
        Func<uint> abiVersion,
        Func<ulong, SharedTrafficScale> scaleTraffic)
    {
        try
        {
            if (abiVersion() != 1)
            {
                return Fallback(nativeText, SharedCoreTrafficShadowStatus.AbiMismatch);
            }

            var sharedText = SharedCoreTrafficDisplayFormatter.Format(
                scaleTraffic(bytes),
                format);
            return string.Equals(sharedText, nativeText, StringComparison.Ordinal)
                ? new SharedCoreTrafficRouteResult(
                    sharedText,
                    SharedCoreTrafficRouteSource.SharedPrimary,
                    SharedCoreTrafficShadowStatus.Matched)
                : Fallback(nativeText, SharedCoreTrafficShadowStatus.Mismatch);
        }
        catch (Exception exception) when (IsNativeCallFailure(exception))
        {
            return Fallback(nativeText, SharedCoreTrafficShadowStatus.NativeCallFailed);
        }
        catch (InvalidOperationException)
        {
            return Fallback(nativeText, SharedCoreTrafficShadowStatus.UnexpectedResult);
        }
        catch (Exception)
        {
            return Fallback(nativeText, SharedCoreTrafficShadowStatus.UnknownFailure);
        }
    }

    public static SharedCoreTrafficLazyRouteResult RouteLazy(
        ulong bytes,
        Func<string> nativeFallback,
        SharedCoreTrafficFormat format)
    {
        return RouteLazy(
            bytes,
            nativeFallback,
            format,
            () => MihomoMeterSharedCore.AbiVersion,
            MihomoMeterSharedCore.ScaleTraffic);
    }

    internal static SharedCoreTrafficLazyRouteResult RouteLazy(
        ulong bytes,
        Func<string> nativeFallback,
        SharedCoreTrafficFormat format,
        Func<uint> abiVersion,
        Func<ulong, SharedTrafficScale> scaleTraffic)
    {
        SharedCoreTrafficRouteStatus fallbackStatus;
        try
        {
            if (abiVersion() != 1)
            {
                fallbackStatus = SharedCoreTrafficRouteStatus.AbiMismatch;
            }
            else
            {
                var sharedText = SharedCoreTrafficDisplayFormatter.Format(
                    scaleTraffic(bytes),
                    format);
                return new SharedCoreTrafficLazyRouteResult(
                    sharedText,
                    SharedCoreTrafficRouteSource.SharedPrimary,
                    SharedCoreTrafficRouteStatus.Succeeded);
            }
        }
        catch (Exception exception) when (IsNativeCallFailure(exception))
        {
            fallbackStatus = SharedCoreTrafficRouteStatus.NativeCallFailed;
        }
        catch (InvalidOperationException)
        {
            fallbackStatus = SharedCoreTrafficRouteStatus.UnexpectedResult;
        }
        catch (Exception)
        {
            fallbackStatus = SharedCoreTrafficRouteStatus.UnknownFailure;
        }

        return LazyFallback(nativeFallback, fallbackStatus);
    }

    private static SharedCoreTrafficRouteResult Fallback(
        string nativeText,
        SharedCoreTrafficShadowStatus status)
    {
        return new SharedCoreTrafficRouteResult(
            nativeText,
            SharedCoreTrafficRouteSource.NativeFallback,
            status);
    }

    private static SharedCoreTrafficLazyRouteResult LazyFallback(
        Func<string> nativeFallback,
        SharedCoreTrafficRouteStatus status)
    {
        return new SharedCoreTrafficLazyRouteResult(
            nativeFallback(),
            SharedCoreTrafficRouteSource.NativeFallback,
            status);
    }

    private static bool IsNativeCallFailure(Exception exception)
    {
        return exception is DllNotFoundException
            or EntryPointNotFoundException
            or BadImageFormatException
            or SEHException;
    }
}

File: Sources/Application/SharedCoreTrafficRouter.swift
enum SharedCoreTrafficRouteSource: String, Hashable, Sendable {
  case sharedPrimary = "shared_primary"
  case nativeFallback = "native_fallback"
}

enum SharedCoreTrafficRouteStatus: String, Hashable, Sendable {
  case matched
  case succeeded
  case abiMismatch = "abi_mismatch"
  case nativeCallFailed = "native_call_failed"
  case unexpectedResult = "unexpected_result"
  case mismatch
  case unknownFailure = "unknown_failure"
}

struct SharedCoreTrafficRouteResult: Equatable, Sendable {
  let text: String
  let source: SharedCoreTrafficRouteSource
  let status: SharedCoreTrafficShadowStatus
}

struct SharedCoreTrafficLazyRouteResult: Equatable, Sendable {
  let text: String
  let source: SharedCoreTrafficRouteSource
  let status: SharedCoreTrafficRouteStatus
}

enum SharedCoreTrafficRouter {
  static func route(
    bytes: UInt64,
    nativeText: String,
    format: SharedCoreTrafficFormat,
    scaleTraffic: (UInt64) throws -> SharedTrafficScale =
      MihomoMeterSharedCoreAdapter.scaleTraffic(bytes:)
  ) -> SharedCoreTrafficRouteResult {
    do {
      let scale = try scaleTraffic(bytes)
      let sharedText = try SharedCoreTrafficDisplayFormatter.string(
        from: scale,
        format: format
      )
      guard sharedText == nativeText else {
        return fallback(nativeText: nativeText, status: .mismatch)
      }
      return SharedCoreTrafficRouteResult(
        text: sharedText,
        source: .sharedPrimary,
        status: .matched
      )
    } catch let error as SharedCoreAdapterError {
      return fallback(nativeText: nativeText, status: status(for: error))
    } catch is SharedCoreTrafficDisplayFormatter.Error {
      return fallback(nativeText: nativeText, status: .unexpectedResult)
    } catch {
      return fallback(nativeText: nativeText, status: .unknownFailure)
    }
  }

  static func routeLazy(
    bytes: UInt64,
    nativeFallback: () -> String,
    format: SharedCoreTrafficFormat,
    abiVersion: () -> UInt32 = { MihomoMeterSharedCoreAdapter.abiVersion },
    scaleTraffic: (UInt64) throws -> SharedTrafficScale =
      MihomoMeterSharedCoreAdapter.scaleTraffic(bytes:)
  ) -> SharedCoreTrafficLazyRouteResult {
    guard abiVersion() == 1 else {
      return lazyFallback(nativeFallback: nativeFallback, status: .abiMismatch)
    }

    do {
      let scale = try scaleTraffic(bytes)
      let sharedText = try SharedCoreTrafficDisplayFormatter.string(
        from: scale,
        format: format
      )
      return SharedCoreTrafficLazyRouteResult(
        text: sharedText,
        source: .sharedPrimary,
        status: .succeeded
      )
    } catch let error as SharedCoreAdapterError {
      return lazyFallback(
        nativeFallback: nativeFallback,
        status: lazyStatus(for: error)
      )
    } catch is SharedCoreTrafficDisplayFormatter.Error {
      return lazyFallback(nativeFallback: nativeFallback, status: .unexpectedResult)
    } catch {
      return lazyFallback(nativeFallback: nativeFallback, status: .unknownFailure)
    }
  }

  private static func fallback(
    nativeText: String,
    status: SharedCoreTrafficShadowStatus
  ) -> SharedCoreTrafficRouteResult {
    SharedCoreTrafficRouteResult(
      text: nativeText,
      source: .nativeFallback,
      status: status
    )
  }

  private static func lazyFallback(
    nativeFallback: () -> String,
    status: SharedCoreTrafficRouteStatus
  ) -> SharedCoreTrafficLazyRouteResult {
    SharedCoreTrafficLazyRouteResult(
      text: nativeFallback(),
      source: .nativeFallback,
      status: status
    )
  }

  private static func status(
    for error: SharedCoreAdapterError
  ) -> SharedCoreTrafficShadowStatus {
    switch error {
    case .nativeCallFailed:
      .nativeCallFailed
    case .unsupportedABIVersion:
      .abiMismatch
    case .unsupportedTrafficUnit:
      .unsupportedTrafficUnit
    }
  }

  private static func lazyStatus(
    for error: SharedCoreAdapterError
  ) -> SharedCoreTrafficRouteStatus {
    switch error {
    case .nativeCallFailed:
      .nativeCallFailed
    case .unsupportedABIVersion:
      .abiMismatch
    case .unsupportedTrafficUnit:
      .unexpectedResult
    }
  }
}
